### ISSUE TITLE
synapse gets stuck in a loop with `Invalid prev_events for <event_id>`

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#11802]
If inbound federation traffic backs up in a v1 room, synapse starts spamming the logs with messages like:

```
2022-01-23 00:42:23,593 - synapse.storage.databases.main.event_federation - 1436 - INFO - _process_incoming_pdus_in_room_inner-1197913 - Invalid prev_events for <event_id>
```

this means that rather than dropping excess traffic as it should, it ends up in a tight loop and fails to process incoming traffic from other sources.

**Comment by richvdh** [comment]
see 2b9f741f3 which is patches this on the matrix-org-hotfixes branch

**Comment by richvdh** [comment]
(dates back to https://github.com/matrix-org/synapse/pull/10390)

**Comment by richvdh** [comment]
A workaround is to for all users on the server to leave the room containing the events (or to use the [delete room](https://matrix-org.github.io/synapse/develop/admin_api/rooms.html#version-2-new-version) admin API to shut it down).

**Comment by Cknight70** [comment]
Just applied this patch.
Previously my homeserver.log would get to 2gb in 30 minutes, Now it is only a few megabytes.
Ram usage is cut in half, messages are reliably sending again, thanks.

**Comment by Redmauss** [comment]
I have this same issue, with ```synapse.storage.databases.main.event_federation - 1427 - INFO - _process_incoming_pdus_in_room_inner-4792 - Invalid prev_events for $164289566819600CTQgK:matrix.kiwifarms.net```

being the log message.  I also have been getting connection interrupts and obscene resource usage(100% CPU and maxing out 4GB mem).  The logs were so constant that my server storage was filled up with a 153GB log file.  I fixed this by making logging in `/etc/matrix-synapse/log.yaml` ERROR instead of INFO. 
 
How do I apply the patch mentioned by @richvdh?  Sorry I haven't made any changes to synapse besides the installation and setting up email and turnserver.  Any help would be greatly appreciated.

**Comment by Cknight70** [comment]
There is probably a more graceful way to do this with `patch` but I did the following:
1. I found the location of synapse from the systemd service file `/lib/systemd/system/matrix-synapse` For me this was `/opt/venvs/matrix-synapse/lib/python3.9/site-packages/synapse/storage/databases/main/`
2. create a backup `cp event_federation.py ~`
3. Stop matrix synapse `sudo systemctl stop matrix-synapse`
4. Remove the old version `sudo rm event_federation.py`
5. Download the patched version `sudo wget https://raw.githubusercontent.com/matrix-org/synapse/2b9f741f3a9dee93e9744774b342c35ce60062c4/synapse/storage/databases/main/event_federation.py`
4. You can now start synapse.

**Comment by anoadragon453** [comment]
Fixed in https://github.com/matrix-org/synapse/pull/11806. The fix will be released in Synapse v1.50.2 and v1.51.0rc2.

Thank you all for testing the patch :slightly_smiling_face:
